### ISSUE TITLE
[Feature Request] Add zstd compression

### ISSUE THREAD

**Issue by tiziodcaio** [NeoApplications/Neo-Backup#444]
It is possible to add the zstd or generally more compressions algorithms?

**Comment by machiav3lli** [comment]
This should be possible, but the "why" question is what would interest me.

**Comment by tiziodcaio** [comment]
Ahhah I asked it if is not too complicated
I'm not expert but gzip has not actually the best compression ratio and it can be cool to use others.
I don't if you are not intersted it was only a little idea 😅

**Comment by opusforlife2** [comment]
Is compression/decompression speed a bottleneck for the app? If so, will it benefit from using zstd?

**Comment by tiziodcaio** [comment]
The clue is that is not a problem really, it was a capricious feature request

**Comment by tiziodcaio** [comment]
If you think is only a dumb idea, no problem saying it. I will close the issue without esitations ;-)

**Comment by hg42** [comment]
Most users will prefer fast compression (e.g. gzip level 1) over an optimal but slow compression ratio.
Though, this could be interesting for speed, if the phone can process the compression faster than the reduced baytes need to crawl to a slow storage (e.g. if using an rclone mount or a slow sdcard and also for syncing).
However, the improved ratio is usually not that big...

**Comment by Leojc** [comment]
Looking forward to this feature! zstd is really fast and has a better compression ratio. So it's totally worth it to replace gzip. Also lz4 is very very fast for decompression, which could be a nice alternative. Lz4 could be about 7x faster than gzip as for decompression i think.

**Comment by tsiflimagas** [comment]
[This](https://github.com/NeoApplications/Neo-Backup/issues/355#issuecomment-939472122) comment from Nils is very interesting. It seems that adding support for more compression algorithms could be much more straightforward since it would use a dependency that's already in the app.

**Comment by hg42** [comment]
> I've found this apps that implement lzo and zstd compression for backups:
> https://github.com/XayahSuSuSu/Android-DataBackup

I didn't look into the source, but DataBackup usrs a script from someone else.
I guess the compression is done via command line while NB uses a library.

Currently you can still switch between complete API/library based archiving (including tar, we call it tarapi, vs tarcmd which uses the tar command).

The compression could be moved to the command line.

However, it would not make sense to do
`tarapi -> compression command > encryption via api`

so switching between tarapi and tarcmd would be dropped, which also removes compatibility for older backups (or the old routines still use the api versions)

**Comment by MrEngineerMind** [comment]
Yeah, not being able to restore an older backup because the compression changed would not be a good thing.

**Comment by MrEngineerMind** [comment]
Using "Often" is not very accurate, and even if it was, this type of "compatibility" issue is kind of unique for this type of app.

Image if adobe photoshop came out with a new version in which you can no longer load legacy BMP image files into it anymore.

**Comment by MrEngineerMind** [comment]
"Eventually...."

But what you are proposing is not a gradual thing...as soon as a new version of NB app switches to a different compression, ALL previous backups are unusable, which is not a good thing.

**Comment by MrEngineerMind** [comment]
The only practical solution is to allow the user to specify the compression method when doing a backup, and have NB auto-detect the compression method of a backup when restoring.

**Comment by MrEngineerMind** [comment]
Being a programmer, I know there would be one-time effort to add support for a second compression method. But that effort would be a fraction of the effort required to do your request of switching to a totally different compression method.

And, once that work is completed, there will be little need to change it for any future versions of NB.

**Comment by MrEngineerMind** [comment]
And that #444 comment is saying that it would be a significant effort to use that other compression method in NB.

**Comment by hg42** [comment]
> The only practical solution is to allow the user to specify the compression method when doing a backup, and have NB auto-detect the compression method of a backup when restoring.

yes, that's the natural thing to do

> Not that practical when you consider that the entire maintenance burden for the alternative code paths is on the developer(s). Totally up to them if they want to spend the time and effort for that.

right, it needs some restructuring, that's the biggest reason for now.
And there are more important things to do. The advantage is too low compared to other things.

The maintenance is not really a problem, when the libs are matured (they just work) and the code is modularized. Up to now compression is integrated with some conditionals and this is not the way to go, if we have multiple compression methods.
E.g. the autodetection doesn't exist, it's "if compressed use .gz and do gzip".
This needs to use the stored compression method or the file extension instead.
And what to do, if they do not match?
A user could recompress it with a different algorithm, which is kind of reasonable from my POV.
The usual (so called professional) approach is to forbid users to manipulate the managed data, but I would like support that (with obvious limitations). Especially for backups, there should be robust strategies.
As a conclusion, I would prefer the file extension.
This has to be discussed between developers.

**Comment by hg42** [comment]
> Being a programmer

nice, so you could add it? :-)

> I know there would be one-time effort to add support for a second compression method

correct, it basically making it modular

supporting more methods would be simple (given a library that supports the same Interface or at least similar enough, in this case it needs to stream the data)

> And, once that work is completed, there will be little need to change it for any future versions of NB

right, here the maturity kicks in. If the lib isn't mature and has bugs, like crashing for certain data, it would create maintenance and even worse, the backup could be unusable despite it was compressed successfully.

This means, I would not like to add any new algorithm, until I have a feeling, that it's really ready for important data.
Note, that many purposes for compression are not mission critical. But backup is (at least from my POV).
(Note, I'm not the main developer)

**Comment by MrEngineerMind** [comment]
"This means, I would not like to add any new algorithm, until I have a feeling, that it's really ready for important data."

I agree.

**Comment by hg42** [comment]
the #444 comment is about a command line tools solution

I wanted to say that even if it's easy for DataBackup, it's another case for NB.

some of my thoughts:

* adding compression to the tar.sh script would indeed be easy
* using a shell executable would either need to add this to NB
* up to now we rejected everything that needs binaries, that are not part of toybox
* using an external executable (e.g. from Termux) would be possible
* preferences for the compress/decompress commands would be enough
* autodetection would still be necessary
* how to handle it, when the user would change those commands, but there are old backups created with other commands?
* handling all this would be much easier if we had plugins
* because plugins are definitely on my plan (but a rather big effort), I tend to postpone all things that could be solved with plugins, at least if the work would be thrown away
* modularization would also help the plugins, so this work would still be useful

**Comment by hg42** [comment]
anyone interested in this, you could also try some things to prove, if the advantage is as big as you hope:

* the stream is
`root file system -> tar -> compression -> encryption -> SAF storage`
* disable compression and compare the speed and size
* do this for different compression levels, less data also means more speed on SAF storage (or even remote)
* note, any comparison on local file systems isn't  helpful. The comparision needs to take SAF into account
* that means faster compression doesn't help much, if the stream is slow at the end

for wizards:
* you can simulate part of the stream by using rclone to mount e.g. an sdcard directory on internal storage, then use tar to compress a directory and stream it though a compressor command and to a file on that mount
* report the commands, the execution times,the sizes, and the rclone mounts
* you may also try remote directories (ssh, gdrive, etc.)

**Comment by hg42** [comment]
> "This means, I would not like to add any new algorithm, until I have a feeling, that it's really ready for important data."

at least, this would be valid for a built in compression method, it is like a suggestion.
A configurable command line or a plugin, that is from an external repository etc. would be their responsibility.

**Comment by murlakatamenka** [comment]
> This should be possible, but the "why" question is what would interest me.

1. It beats old gzip bacisally everywhere (compression/decompression speed/time, compression ratio, linear scaling of compression depending on selected level.

https://morotti.github.io/lzbench-web -> section Evolution:

> We're in the 3rd millenimum and there was surprisingly little progress in general compression in the past decades. deflate, lzma and lzo are from the 90's, the origin of lz compression traces back to at least the 70's.
>
> Actually, it's not true that nothing happened. Google and Facebook have people working on compression, they have a lot of data and a ton to gain by shaving off a few percents here and there.
>
> Facebook in particular has hired the top compression research scientist and rolled 2 compressors based on a novel compression approach that is doing wonder. That could very well be the biggest advance in computing in the last decade.
>
> See [zstd](https://github.com/facebook/zstd) (medium) and [lz4](https://github.com/lz4/lz4) (fast):
>
>    - zstd blows deflate out of the water, achieving a better compression ratio than gzip while being multiple times faster to compress.
>    - lz4 blows lzo and google snappy by all metrics, by a fair margin.
>
> Better yet, they come with a wide range of compression levels that can adjust speed/ratio almost linearly. The slower end pushes against the other slow algorithms, while the fast end pushes against the other faster algorithms. It's incredibly friendly as a developer or a user. All it takes is a single algorithm to support (zstd) with a single tunable setting (1 to 20) and it's possible to accurately tradeoff speed for compression. It's unprecedented.
>
> Of course one could say that gzip already offerred tunable compression levels (1-9) however it doesn't cover a remotely comparable range of speed/ratio. Not to mention that the upper half is hardly useful, it's already slow and making it slower for little benefit.

---

This is from bench made 6 years ago (!), and it surely improved since that time.

2. zstd is in the Linux kernel, and kernel devs don't do random crap into it

    zstd can be used to compress the kernel itself and its modules, zram/zswap, for transparent filesystem compression (BTRFS)

---

  This alone is the ultimate seal of approval, Linux kernel is no joke 

3. It's basically everywhere now?
  
    See https://en.wikipedia.org/wiki/Zstd#Usage

    A few examples:
  
    - [Arch Linux compresses its packages with zstd since 2019](https://archlinux.org/news/now-using-zstandard-instead-of-xz-for-package-compression/)
    - [Fedora uses zstd for their RPM](https://fedoraproject.org/wiki/Changes/Switch_RPMs_to_zstd_compression#Switch_RPMs_to_zstd_compression)

    Repos of Linux packages are no joke too, this scenario is pretty similar to backup one:

    - fast compression/decompression -> job done faster, happy user, less battery used
    - smaller size -> less storage/bandwidth used

---

The fact that this issues is opened for 2 years is quite disappointing imo.

**Comment by murlakatamenka** [comment]
As for implementation, I see that compression library used [common-compress](https://mvnrepository.com/artifact/org.apache.commons/commons-compress) supports zstd, so I don't really get why it should be hard to use another compressor:

`tar | gzip | encryption -> org.jdoe.app.tar.gz`
->
`tar | zstd | encryption -> org.jdoe.app.tar.zst`

Add compression + level

```rs
enum Compression {
    Gzip(u8)
    Zstd(u8)
}
```

and expose those via GUI, which at the moment exposes only level because there is only gzip. Voilà? While simplified, isn't it right?

---

@hg42 

> autodetection would still be necessary

that's not a problem, zstd-compressed stream has its own magic number, see
https://datatracker.ietf.org/doc/html/rfc8878#section-3.1.1-3.2

> Magic_Number:
    4 bytes, little-endian format. Value: `0xFD2FB528`

---

> how to handle it, when the user would change those commands, but there are old backups created with other commands?

I'd say this is overthinking. If a user is smart and advanced, he either won't do it, or won't recompress tar.gz into tar.xz and will expect NB to work just like nothing happened.

**Comment by hg42** [comment]
> 1. It beats old gzip bacisally everywhere (compression/decompression speed/time, compression ratio, linear scaling of compression depending on selected level).

the memory consumption would be an important part. Speed often comes from using more memory.
* there are several backups running in parallel
* phones have less memory than workstations and servers

**Comment by hg42** [comment]
> The fact that this issues is opened for 2 years is quite disappointing imo.

well, pull requests are probably welcome, when well done

**Comment by hg42** [comment]
> > autodetection would still be necessary
> that's not a problem

in the sense that it's solved, yes, but I didn't say it's a problem, I meant, it's work to be done.

All I do, is gathering info, to have a picture how much work it is in relation to the gain.

What to do with the info, is a thing of priority lists,
or enthusiastic developers creating pull requests.

From my current POV it's work, that some other developer can do, exactly because it's fairly simple.
I concentrate on more complicated things.

According to compression, I use gzip level 1 and other items on my personal todo list will probably help much more with my backups.

**Comment by murlakatamenka** [comment]
> > 1. It beats old gzip bacisally everywhere (compression/decompression speed/time, compression ratio, linear scaling of compression depending on selected level).
> 
> the memory consumption would be an important part. Speed often comes from using more memory.
> 
>     * there are several backups running in parallel
> 
>     * phones have less memory than workstations and servers

True, I didn't mention memory usage anywhere, but again, zstd  is flexible and allows controlling both memory and threads. 

For example, memory tunable mainly via [`windowLog`](https://man.archlinux.org/man/zstd.1.en#ADVANCED_COMPRESSION_OPTIONS)

For threads there are `-T/--threads` and `--single-thread` arguments, the latter is specifically tailored for low-memory scenarios:

> `--single-thread`: Use a single thread for both I/O and compression. As compression is serialized with I/O, this can be slightly slower. Single-thread mode features significantly lower memory usage, which can be useful for systems with limited amount of memory, such as 32-bit systems.

So both options are available: several parallel backups with single thread per worker and sequential backup where single worker uses multiple threads to compress faster; both with respect to a memory limit.

**Comment by hg42** [comment]
I thought of some benchmark on a mid range phone :-)

But at the end it doesn't matter, it's on a todo list but not priority

**Comment by hg42** [comment]
> For threads there are `-T/--threads` and `--single-thread` arguments, the latter is specifically tailored for low-memory scenarios

this doesn't help, because multiple backups are running in parallel, we would probably use single-thread, anyways

The main question is how much memory zstd uses.
Benchmarks for a single compression don't matter, if the memory consumption is high and it starts to invalidate pages.
So a fast zstd with high memory consumption could be (maybe much) slower than a slower gzip with low memory demands.

**Comment by hg42** [comment]
and if memory consumption would be limited by an option, would it still be faster?

**Comment by FrozzDay** [comment]
I don't have any scientific tests, but my phone, which has 4GB of RAM, can handle the `tar --zstd` operation in 13 seconds and the `tar --gzip` operation in 40 seconds.

**Comment by hg42** [comment]
it's obvious, that a single compression works faster. But as I said, speed AND compression ratio are both a trade with memory consumption.

So, to compare, you need to start 8 (or the number of cores) archiving processes at once for each.

The result could be totally different.

**Comment by hg42** [comment]
this tells us, that zstd should also be good for multithreading:

https://engineering.fb.com/2016/08/31/core-infra/smaller-and-faster-data-compression-with-zstandard/

unless you choose a higher level:

https://jothiprasath.com/blog/gzip-vs-zstd/
see level 19

the meta document tells us, that it can also use terabytes.
Looks like level 3 is a good balance. (well, I already hear people complaining about never ending backups, if they do not keep the default)

**Comment by FrozzDay** [comment]
I tried adding zstd in this app, I think it's quite stable. here is the repo [NeoBackup + zstd](https://github.com/FrozzDay/Neo-Backup)

**Comment by hg42** [comment]
> I tried adding zstd in this app, I think it's quite stable. here is the repo [NeoBackup + zstd](https://github.com/FrozzDay/Neo-Backup)

from a quick check, this looks good.

**Comment by hg42** [comment]
I will create a pumpkin apk soon...

as far as I see compressionLevel is still 0...9, I guess that's ok for now... given that higher levels are slow
I'm currently testing some levels
level 1 created 15 GB in 10 min,
level 3 created 15 GB in 13 min ...
(need to check gzip, I think it also created 15 GB, though maybe it took more time...)

**Comment by hg42** [comment]
I also think gz / zstd should be gzip and zstd (or gz and zst, I would prefer the first), so there is type, extension

**Comment by hg42** [comment]
here my test apk:

https://t.me/neo_backup/53944/56653

**Comment by FrozzDay** [comment]
> or be pragmatic, using file extension? "gz" and "zst" ?
> but if more types are added...

i think that should be enough? as each compression supported by Commons Compress has their own unique format

**Comment by hg42** [comment]
my thought was, that gzip/zstd/whatelse would be more clear than gz, zst.

The names derived from the file extensions are not really standard, e.g. Windows usually uses 3 letters, while others often use more e.g. .zstd is also common.

But I decided to take the short name for now.

At some point compression (and other output transformations) should probably be moved into modules.
Then alternative names can be registered by those modules.

**Comment by hg42** [comment]
I'll close this, because it's implemented by @FrozzDay (thanks again) in #856 .

I'm using it myself, it seems to work well and the code is clean.
Be careful with the compression levels. I think higher levels take much more memory and then it probably slows down. You should probably find the sweet spot for your situation. Though, with lower levels it's a no-brainer.

**Comment by hg42** [comment]
what is good for normal computers is not automatically the same for smartphones and similar devices.

It's a tradeoff...

higher compression correlates with more RAM, more CPU usage, less size = less storage space necessary and faster storage or transfer.

So there is a sweet spot for every combination of
* CPU (cores, architecture, speed)
* RAM (size, speed)
* storage/transfer (space, speed)

It's probably much different for phones compared to servers and PCs.
Servers and PCs usually have more (and faster!) RAM.
CPUs are often more powerful (more cores, more speed, better support for compression).
Storage is usually much faster (SSDs vs. sdcard, or internal flash).
Phones are first optimized for low power, PCs/Servers are usually first optimized for speed.

Another very important difference:
**NB runs several compressions at once** (up to 8 or better the number of cores).
So RAM, CPU and storage are much more constrained for each compression.
That's why statistics about a single compression are not comparable at all.

zstd is probably still better than gzip, because it's more efficient (because it's a much newer algorithm).
This implies gzip wasn't optimal. This seems this is true, but it was optimized for tighter condition, I guess it is still usable in very low RAM or CPU situations.

The best level is probably different for each user and the particular situation.
E.g. running remote backups will prefer higher compression rates, but only until RAM becomes tight and the system starts swapping.
Even swapping can be very different... It's no problem to swap out or even drop sleeping apps. But it's expensive if these apps are started

For me the differences between zstd and gzip are not very big, though I probably have not found the sweet spot, yet.

I would like some reports from users for real situations (please add the conditions).

**Comment by cadexI9wUR7** [comment]
> the #444 comment is about a command line tools solution
> 
> I wanted to say that even if it's easy for DataBackup, it's another case for NB.
> 
I hope you won't mind,
can you point out the pro & con of Neo-Backup against Android-DataBackup？
Apart from the UI, it seemed that both software does the same trick, but nothing online did a comparison or explained the different magics used
Thanks

**Comment by hg42** [comment]
not the right place for this, but...
I would say we do not compare NB to anything, we just try to do what we think is good in a speed that results from the spare time we have.

Basically both apps do the same, saving the different data parts of each selected app.

When I last looked, Data Backup was a frontend to the cbackup shell script (I think it still is).
This has advantages but also disadvantages.

I think (from theory, never checked) that a script runs without interruption by the system, while an app can be put to sleep in certain cases etc.
But, becasue of this NB uses a work queue for it's jobs. There is still a problem, if a single archive takes very long (e.g. in the range of more than 10 minutes).

The script will just be killed by the shutdown.
NB continues it's work, if the phone is rebooted, because it uses the Android WorkManager to queue it's job.

That's also a disadvantage for users that don't expect this (and cannot read :-).

From memory, NB had more data parts, but I see they catched up with it.

Not sure what is meant with "100% Data Integrity", either they have a trick or it's only a wish. 100% needs freezing the file systems (and all at once, which is usually impossible)...
